src/cellrank/estimators/terminal_states/_gpcca.py

```diff
@@ -532,6 +534,150 @@ def set_initial_states(
         )
         return self
 
+    def get_tsi(self, n_macrostates: int, terminal_states: List[str], cluster_key: str) -> pd.DataFrame:
+        """Compute terminal state identificiation (TSI).
+
+        Parameters
+        ----------
+        n_macrostates
+            Maximum number of macrostates to consider.
+        terminal_states
+            List of terminal states.
+        cluster_key
+            Key in :attr:`~anndata.AnnData.obs` defining cluster labels including terminal states.
+
+        Returns
+        -------
+        Returns TSI as a Pandas DataFrame and adds the class attribute :attr:`tsi`.
+        """
+        macrostates = {}
+        for n_states in range(1, n_macrostates):
+            self.compute_macrostates(n_states=n_states, cluster_key=cluster_key)
+            macrostates[n_states] = self._macrostates.assignment.cat.categories
+
+        max_terminal_states = len(terminal_states)
+
+        tsi_df = {
+            "Number of macrostates": [],
+            "Identified terminal states": [],
+            "Optimal identification": [],
+        }
+        for n_states, states in macrostates.items():
+            n_terminal_states = (
+                states.str.replace(r"(_).*", "", regex=True).drop_duplicates().isin(terminal_states).sum()
+            )
+            tsi_df["Number of macrostates"].append(n_states)
+            tsi_df["Identified terminal states"].append(n_terminal_states)
+
+            if n_states <= max_terminal_states:
+                tsi_df["Optimal identification"].append(n_states)
+            else:
+                tsi_df["Optimal identification"].append(max_terminal_states)
+
+        tsi_df = pd.DataFrame(tsi_df)
+        self.tsi = tsi_df
+
+        return tsi_df
+
+    def get_tsi_score(self) -> float:
+        """Compute TSI score."""
+        if not hasattr(self, "tsi"):
+            raise RuntimeError("Compute TSI with `get_tsi` first.")
+
+        optimal_score = self.tsi["Optimal identification"].sum()
+
+        return self.tsi["Identified terminal states"].sum() / optimal_score
+
+    def plot_tsi(
+        self,
+        tsi_df: pd.DataFrame,
+        fname: Optional[PosixPath] = None,
+        x_offset: Optional[Tuple[float, float]] = None,
+        y_offset: Optional[Tuple[float, float]] = None,
+        **kwargs: Any,
+    ):
+        """Plot terminal state identificiation (TSI).
+
+        Parameters
+        ----------
+        tsi_df
+            Pre-computed TSI DataFrame.
+        fname
```

Comment:
We usually call it `save` and also include `dpi` + `figsize` in general (see other plotting methods, there's a `docrep` stuff for this as well), please use this.

Reply:
Yes, I know. I honestly don't like it because it doesn't convey what the argument is meant for; the corresponding argument in `savefig` is also called `fname`. I changed it, though.